runtime-modules/referendum/src/lib.rs

```diff
@@ -0,0 +1,814 @@
+            }
+
+            // is there draw in the last winning place?
+            if winning_order[(target_count as usize) - 1].1
+                == winning_order[target_count as usize].1
+            {
+                let mut draw_end_index = target_count as usize;
+                while voted_options_count > draw_end_index + 1
+                    && winning_order[draw_end_index].1 == winning_order[draw_end_index + 1].1
+                {
+                    draw_end_index += 1;
+                }
+
+                return ReferendumResult::ExtraWinners(
+                    winning_order[..(draw_end_index as usize + 1)].to_vec(),
+                );
+            }
+
+            // return winning options
+            ReferendumResult::Winners(winning_order[..(target_count as usize)].to_vec())
+        }
+
+        // calculate votes and select winner(s)
+        let referendum_result = calculate_votes::<T, I>(revealing_stage);
+
+        // reset referendum state
+        Self::reset_referendum(&referendum_result);
+
+        // return winning option
+        referendum_result
+    }
+
+    fn reset_referendum(last_cycle_result: &ReferendumResult<u64, T::VotePower>) {
+        Stage::<T, I>::put(ReferendumStage::Inactive);
+        CurrentCycle::<I>::put(CurrentCycle::<I>::get() + 1);
+        PreviousCycleResult::<T, I>::put(last_cycle_result);
+    }
+
+    /// Can return error when stake fails to lock
+    fn vote(
+        account_id: &<T as system::Trait>::AccountId,
+        commitment: &T::Hash,
+        balance: &Balance<T, I>,
+    ) -> Result<(), Error<T, I>> {
+        // lock stake amount
+        T::Currency::set_lock(
+            T::LockId::get(),
+            account_id,
+            *balance,
+            WithdrawReason::Transfer.into(),
+        );
+
+        // store vote
+        Votes::<T, I>::insert(
+            account_id,
+            SealedVote {
+                commitment: *commitment,
+                balance: *balance,
+                cycle_id: CurrentCycle::<I>::get(),
+                vote_for: None,
+            },
+        );
+
+        Ok(())
+    }
+
+    fn reveal_vote(
+        account_id: &<T as system::Trait>::AccountId,
+        option_index: &u64,
+        sealed_vote: &SealedVote<T::Hash, Balance<T, I>>,
+    ) -> Result<(), Error<T, I>> {
+        let distribute_vote =
+            |stage_data: &mut ReferendumStageRevealing<T::BlockNumber, T::VotePower>| {
+                // calculate vote power
+                let vote_power = T::caclulate_vote_power(account_id, &sealed_vote.balance);
+                stage_data.revealed_votes[*option_index as usize] += vote_power;
+            };
+
+        // store revealed vote
+        Stage::<T, I>::mutate(|stage| {
+            match stage {
+                ReferendumStage::Revealing(stage_data) => distribute_vote(stage_data),
+                _ => panic!("Invalid state"), // will never happen
```

Comment:
This is avoidable. Recall that `reveal_vote` is invoked after `can_reveal_vote` has succeeded in the same code path, so you can provide required inputs for this function here with static guarantees about `stage_data` being available.

In fact, the problem is a bit deeper, look at this line above

`stage_data.revealed_votes[*option_index as usize] += vote_power;`

this is not safe, `option_index` can be out of bounds, again despite the fact that you have in fact earlier in the code path ensured that the `option_index` is valid. I suggest you combine this check for validity of `option_index` with a pure computation of the new `revealed_votes` vector based on `stage_data`.

Reply:
Reworked